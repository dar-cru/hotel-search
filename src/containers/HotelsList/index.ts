export { default } from './HotelsList.connected';
