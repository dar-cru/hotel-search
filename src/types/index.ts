export * from './hotel';
export * from './offer';
export * from './property';
export * from './rating';
