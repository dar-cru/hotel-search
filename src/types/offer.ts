export enum PromotionType {
  CAMPAIGN = 'CAMPAIGN',
  MEMBER = 'MEMBER'
}

export enum CancellationType {
  NOT_REFUNDABLE = 'NOT_REFUNDABLE',
  FREE_CANCELLATION = 'FREE_CANCELLATION'
}

export type Offer = {
  promotion: {
    title: string;
    type: PromotionType;
  };
  name: string;
  displayPrice: {
    amount: number;
    currency: string;
  };
  savings?: {
    amount: number;
    currency: string;
  };
  cancellationOption: {
    cancellationType: CancellationType;
  };
};
