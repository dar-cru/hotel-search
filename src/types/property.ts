import { Rating } from './rating';

export type Property = {
  propertyId: string;
  title: string;
  address: string[];
  previewImage: {
    url: string;
    caption: string;
    imageType: string;
  };
  rating: Rating;
};
