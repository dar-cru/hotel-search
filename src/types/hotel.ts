import { Offer } from './offer';
import { Property } from './property';

export type Hotel = {
  id: string;
  property: Property;
  offer: Offer;
};
