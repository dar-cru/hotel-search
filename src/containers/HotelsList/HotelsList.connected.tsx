import { useEffect, useState } from 'react';
import HotelsList from './HotelsList';
import { getHotels } from '../../services/hotels.api';
import { Hotel } from '../../types';

const Connected = () => {
  const [hotels, setHotels] = useState<Hotel[] | undefined>();
  const [error, setError] = useState<string | undefined>();

  useEffect(() => {
    const fetchHotels = async () => {
      try {
        setError('');
        const hotelData = (await getHotels()) as Hotel[];
        setHotels(hotelData);
      } catch (e) {
        setError(e.message);
      }
    };

    fetchHotels();
  }, []);

  return (
    <div>{error ? <div style={{ color: 'red' }}>{error}</div> : <>{hotels && <HotelsList hotels={hotels} />}</>}</div>
  );
};

export default Connected;
