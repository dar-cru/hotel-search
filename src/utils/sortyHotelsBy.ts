import { Hotel } from '../types/hotel';

export enum SortOrder {
  ASCENDING,
  DESCENDING
}

export enum SortCategory {
  DISPLAY_PRICE
}

export const sortHotelsBy = (hotels: Hotel[], category: SortCategory, order: SortOrder) => {
  return hotels.sort((hotel1, hotel2) => {
    const obj1 = order === SortOrder.ASCENDING ? hotel1 : hotel2;
    const obj2 = order === SortOrder.ASCENDING ? hotel2 : hotel1;

    if (category === SortCategory.DISPLAY_PRICE) {
      return obj1.offer.displayPrice.amount - obj2.offer.displayPrice.amount;
    } else {
      return 0;
    }
  });
};
