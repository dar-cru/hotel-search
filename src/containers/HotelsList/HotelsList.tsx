import { ChangeEvent, useState } from 'react';
import HotelCard from '../../components/HotelCard';
import { Hotel } from '../../types/hotel';
import { sortHotelsBy, SortValues } from '../../utils/sortyHotelsBy';

type HotelsListProps = {
  hotels: Hotel[];
};

const HotelsList = ({ hotels }: HotelsListProps) => {
  const [sortedHotels, setSortedHotels] = useState<Hotel[]>(hotels);

  const onHotelsSort = (event: ChangeEvent<HTMLSelectElement>) => {
    const value = event.target.value;
    const sortedValues = sortHotelsBy([...sortedHotels], value as SortValues);
    setSortedHotels(sortedValues);
  };

  return (
    <>
      <div style={{ display: 'flex' }}>
        <div data-testid="hotel-results-summary" style={{ padding: '10px', width: '50%' }}>
          <b>{sortedHotels?.length}</b> <i>hotels in</i> <b>Sydney</b>
        </div>

        <div style={{ padding: '10px', width: '50%', justifyContent: 'flex-end', display: 'flex' }}>
          <p style={{ margin: 0, paddingRight: '4px' }}>Sort by</p>

          <select data-testid="hotel-sort-selector" defaultValue="default" onChange={onHotelsSort}>
            <option value="default" disabled={true} hidden={true}>
              Default
            </option>
            <option data-testid={SortValues.PRICE_LOW_TO_HIGH} value={SortValues.PRICE_LOW_TO_HIGH}>
              Price low-high
            </option>
            <option data-testid={SortValues.PRICE_HIGH_TO_LOW} value={SortValues.PRICE_HIGH_TO_LOW}>
              Price high-low
            </option>
          </select>
        </div>
      </div>

      {sortedHotels?.map((hotel) => {
        return <HotelCard key={`hotel-${hotel.id}`} property={hotel.property} offer={hotel.offer} />;
      })}
    </>
  );
};

export default HotelsList;
