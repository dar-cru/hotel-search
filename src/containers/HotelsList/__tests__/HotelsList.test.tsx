import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SortValues } from '../../../utils/sortyHotelsBy';
import HotelsList from '../HotelsList';
import { CancellationType, Hotel, PromotionType, RatingType } from '../../../types';

const hotels: Hotel[] = [
  {
    id: 'cxd650nuyo',
    property: {
      propertyId: 'P107801',
      title: 'Courtyard by Marriott Sydney-North Ryde',
      address: ['7-11 Talavera Rd', 'North Ryde'],
      previewImage: {
        url: 'https://unsplash.it/145/125/?random',
        caption: 'Image of Courtyard by Marriott Sydney-North Ryde',
        imageType: 'PRIMARY'
      },
      rating: {
        ratingValue: 4.5,
        ratingType: RatingType.SELF
      }
    },
    offer: {
      promotion: {
        title: 'Exclusive Deal',
        type: PromotionType.MEMBER
      },
      name: 'Deluxe Balcony Room',
      displayPrice: {
        amount: 329.0,
        currency: 'AUD'
      },
      savings: {
        amount: 30.0,
        currency: 'AUD'
      },
      cancellationOption: {
        cancellationType: CancellationType.NOT_REFUNDABLE
      }
    }
  },
  {
    id: 'mesq6mggyn',
    property: {
      propertyId: 'P107802',
      title: 'Primus Hotel Sydney',
      address: ['339 Pitt St', 'Sydney'],
      previewImage: {
        url: 'https://unsplash.it/145/126/?random',
        caption: 'Image of Primus Hotel Sydney',
        imageType: 'PRIMARY'
      },
      rating: {
        ratingValue: 5,
        ratingType: RatingType.STAR
      }
    },
    offer: {
      promotion: {
        title: 'Exclusive Deal',
        type: PromotionType.CAMPAIGN
      },
      name: 'Deluxe King',
      displayPrice: {
        amount: 375.0,
        currency: 'AUD'
      },
      savings: {
        amount: 28.0,
        currency: 'AUD'
      },
      cancellationOption: {
        cancellationType: CancellationType.FREE_CANCELLATION
      }
    }
  }
];

describe('HotelsList container', () => {
  it('renders the hotels list in the order of API response', () => {
    const tree: RenderResult = render(<HotelsList hotels={hotels} />);

    const { container } = tree;
    expect(container).toMatchSnapshot();

    const hotelResultsSummary = screen.getByTestId('hotel-results-summary');
    const hotelCardItems = screen.getAllByTestId('hotel-name');

    expect(hotelResultsSummary?.textContent).toEqual('2 hotels in Sydney');
    expect(hotelCardItems.length).toBe(2);
    expect(hotelCardItems[0].textContent).toEqual('Courtyard by Marriott Sydney-North Ryde');
    expect(hotelCardItems[1].textContent).toEqual('Primus Hotel Sydney');
  });

  it('should sort the hotels by price - high to low, then low to high', () => {
    render(<HotelsList hotels={hotels} />);

    // ACT: Trigger a select event with value - high to low
    userEvent.selectOptions(screen.getByTestId('hotel-sort-selector'), [SortValues.PRICE_HIGH_TO_LOW]);

    let hotelCardItems = screen.getAllByTestId('hotel-name');
    expect(hotelCardItems.length).toBe(2);
    expect(hotelCardItems[0].textContent).toEqual('Primus Hotel Sydney');
    expect(hotelCardItems[1].textContent).toEqual('Courtyard by Marriott Sydney-North Ryde');

    // ACT: Retrigger the selector event but with a different value - low to high
    userEvent.selectOptions(screen.getByTestId('hotel-sort-selector'), [SortValues.PRICE_LOW_TO_HIGH]);

    hotelCardItems = screen.getAllByTestId('hotel-name');
    expect(hotelCardItems.length).toBe(2);
    expect(hotelCardItems[0].textContent).toEqual('Courtyard by Marriott Sydney-North Ryde');
    expect(hotelCardItems[1].textContent).toEqual('Primus Hotel Sydney');
  });
});
