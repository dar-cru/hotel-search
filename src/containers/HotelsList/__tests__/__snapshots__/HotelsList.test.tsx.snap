// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelsList container renders the hotels list in the order of API response 1`] = `
.c1 {
  position: relative;
  width: 242px;
  height: 150px;
  margin: 5px 10px 5px 0;
}

.c2 {
  position: absolute;
  width: 100%;
  height: 100%;
}

.c3 {
  position: absolute;
  background-color: white;
  padding: 6px;
  top: 12px;
  left: 0;
  box-sizing: border-box;
  color: red;
  margin: 0;
  font-size: 0.9em;
  font-weight: 500;
}

.c7 {
  display: inline-block;
  font-size: 1.5rem;
  font-family: Times;
  line-height: 1;
}

.c7:before {
  content: '●●●●●';
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  background: linear-gradient(90deg,gold 90%,grey 90%);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

.c13 {
  display: inline-block;
  font-size: 1rem;
  font-family: Times;
  line-height: 1;
}

.c13:before {
  content: '★★★★★';
  -webkit-letter-spacing: 3px;
  -moz-letter-spacing: 3px;
  -ms-letter-spacing: 3px;
  letter-spacing: 3px;
  background: linear-gradient(90deg,gold 100%,grey 100%);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: relative;
  box-sizing: border-box;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  box-sizing: border-box;
}

.c4 {
  margin: 0;
  background-color: #cbcaca;
  border: 0;
  height: 1px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 70%;
  padding-top: 20px;
}

.c6 {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  margin: 0;
  padding-right: 8px;
}

.c9 {
  color: #999999;
  margin: 0;
  font-size: 0.8rem;
}

.c10 {
  position: absolute;
  bottom: 18px;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #e40000;
  font-size: 0.8rem;
  margin: 0;
}

.c11 {
  font-size: 2rem;
  color: #3f3f3f;
  font-weight: 200;
  padding: 4px 0;
  margin: 0;
}

.c11:before {
  content: '$';
  font-size: 1.15rem;
  vertical-align: 60%;
  padding-right: 2px;
}

.c12 {
  font-size: 1rem;
  color: #e40000;
  margin: 0;
}

.c12:after {
  content: '~';
  font-size: 0.8rem;
  vertical-align: 25%;
  padding-left: 1px;
}

.c14 {
  position: absolute;
  margin: 0;
  bottom: 8px;
  color: #35a509;
  font-size: 0.8rem;
}

<div>
  <div
    style="display: flex;"
  >
    <div
      data-testid="hotel-results-summary"
      style="padding: 10px; width: 50%;"
    >
      <b>
        2
      </b>
       
      <i>
        hotels in
      </i>
       
      <b>
        Sydney
      </b>
    </div>
    <div
      style="padding: 10px; width: 50%; justify-content: flex-end; display: flex;"
    >
      <p
        style="margin: 0px; padding-right: 4px;"
      >
        Sort by
      </p>
      <select
        data-testid="hotel-sort-selector"
      >
        <option
          disabled=""
          hidden=""
          selected=""
          value="default"
        >
          Default
        </option>
        <option
          data-testid="PRICE_LOW_HIGH"
          value="PRICE_LOW_HIGH"
        >
          Price low-high
        </option>
        <option
          data-testid="PRICE_HIGH_LOW"
          value="PRICE_HIGH_LOW"
        >
          Price high-low
        </option>
      </select>
    </div>
  </div>
  <div
    class="c0"
    data-testid="hotel-card-item"
    direction="row"
    style="padding-bottom: 5px;"
  >
    <div
      class="c1"
    >
      <img
        alt="Image of Courtyard by Marriott Sydney-North Ryde"
        class="c2"
        src="https://unsplash.it/145/125/?random"
      />
      <p
        class="c3"
      >
        Exclusive Deal
      </p>
    </div>
    <div
      style="width: 100%; padding: 0px 10px;"
    >
      <hr
        class="c4"
      />
      <div
        class="c5"
      >
        <h2
          class="c6"
          data-testid="hotel-name"
        >
          Courtyard by Marriott Sydney-North Ryde
        </h2>
        <div
          class="c7"
        />
      </div>
      <div
        class="c0"
        direction="row"
        style="min-height: 116px;"
      >
        <div
          class="c8"
          direction="column"
          style="max-height: 70px; padding-top: 4px; width: 50%;"
        >
          <p
            class="c9"
            data-testid="hotel-address-text"
          >
            7-11 Talavera Rd, North Ryde
          </p>
          <p
            class="c10"
          >
            Deluxe Balcony Room
          </p>
        </div>
        <div
          class="c8"
          direction="column"
          style="align-items: flex-end; padding-top: 28px; width: 50%;"
        >
          <p
            class="c9"
            data-testid="hotel-night-total-text"
          >
            1 night total (
            AUD
            )
          </p>
          <p
            class="c11"
            data-testid="hotel-price-text"
          >
            329
          </p>
          <p
            class="c12"
            data-testid="hotel-savings-text"
          >
            Save $
            30
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c0"
    data-testid="hotel-card-item"
    direction="row"
    style="padding-bottom: 5px;"
  >
    <div
      class="c1"
    >
      <img
        alt="Image of Primus Hotel Sydney"
        class="c2"
        src="https://unsplash.it/145/126/?random"
      />
      <p
        class="c3"
      >
        Exclusive Deal
      </p>
    </div>
    <div
      style="width: 100%; padding: 0px 10px;"
    >
      <hr
        class="c4"
      />
      <div
        class="c5"
      >
        <h2
          class="c6"
          data-testid="hotel-name"
        >
          Primus Hotel Sydney
        </h2>
        <div
          class="c13"
        />
      </div>
      <div
        class="c0"
        direction="row"
        style="min-height: 116px;"
      >
        <div
          class="c8"
          direction="column"
          style="max-height: 70px; padding-top: 4px; width: 50%;"
        >
          <p
            class="c9"
            data-testid="hotel-address-text"
          >
            339 Pitt St, Sydney
          </p>
          <p
            class="c10"
          >
            Deluxe King
          </p>
        </div>
        <div
          class="c8"
          direction="column"
          style="align-items: flex-end; padding-top: 28px; width: 50%;"
        >
          <p
            class="c9"
            data-testid="hotel-night-total-text"
          >
            1 night total (
            AUD
            )
          </p>
          <p
            class="c11"
            data-testid="hotel-price-text"
          >
            375
          </p>
          <p
            class="c12"
            data-testid="hotel-savings-text"
          >
            Save $
            28
          </p>
        </div>
      </div>
      <p
        class="c14"
        data-testid="hotel-free-cancellation"
      >
        Free Cancellation
      </p>
    </div>
  </div>
</div>
`;
